### ISSUE TITLE
TextLoader Separator/SeparatorChars

### ISSUE THREAD

**Issue by TomFinley** [dotnet/machinelearning#2041]
Consider these two settings:

https://github.com/dotnet/machinelearning/blob/496e185fdaf8969ffd856c65c42dd259368a242d/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs#L356-L360

They both exist for the same purpose but serve very different endpoints. One exists for the benefit of the command line, one for the benefit of the API. The API one (currently named `SeparatorChars`) should instead become `Separators` (plural pursuant to #2040). The other, currently named `Separator`, can remain named that, but should be invisible to API users since it exists purely for the benefit of the command line users. So, it should become `internal`.

/cc @stephentoub

**Comment by mareklinka** [comment]
This seems rather straightforward, except that `CmdParser.GetArgumentInfo` will have to reflect on both public and private fields, which might potentially lead to unexpected side-effects.

On the other hand, any field still needs to be annotated by the `ArgumentAttribute` attribute, which would indicate that the field was indeed meant to be used as an argument, limiting any confusion.

And there might be other such CMD-only fields, thouse could also be made internal then, right?

**Comment by TomFinley** [comment]
That's a great point @mareklinka ... you have earned your N7 armor well! 😄 Does `CmdParser.GetArgumentInfo` work only over public fields? I hope it works over internal fields as well. But, maybe it does not. Certainly if a field is internal *but* has the attribute, we would like it covered. (Indeed I have some other code in the works that relies on this being the case, so if it's not, then I'll have to fix it anyway!!)

But you're right, I actually don't know... let's test to see if this also needs to be expanded.

https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Commands/SaveDataCommand.cs

    dotnet MML.dll ? savedata

So this gets this info:

```
Help for Command: 'SaveData'
  Aliases: save

Summary:
   Given input data, a loader, and possibly transforms, save the data to a new file as parameterized by a saver.

saver=<name>{<options>}             The data saver to use
outputDataFile=<string>             File to save the data (short form dout)
keepHidden=[+|-]                    Whether to include hidden columns Default value:'-' (short form keep)
loader=<name>{<options>}            The data loader
dataFile=<string>                   The data file (short form data)
outputModelFile=<string>            Model file to save (short form out)
inputModelFile=<string>             Model file to load (short form in)
loadTransforms=[+|-]                Load transforms from model file? (short form loadTrans)
randomSeed=<int>                    Random seed (short form seed)
parallel=<int>                      Desired degree of parallelism in the data pipeline (short form n)
transform[<tag>]=<name>{<options>}  Transform (short form xf)
```

Ah, and if I make that field `internal`, I get this:

```
Help for Command: 'SaveData'
  Aliases: save

Summary:
   Given input data, a loader, and possibly transforms, save the data to a new file as parameterized by a saver.

outputDataFile=<string>             File to save the data (short form dout)
keepHidden=[+|-]                    Whether to include hidden columns Default value:'-' (short form keep)
loader=<name>{<options>}            The data loader
dataFile=<string>                   The data file (short form data)
outputModelFile=<string>            Model file to save (short form out)
inputModelFile=<string>             Model file to load (short form in)
loadTransforms=[+|-]                Load transforms from model file? (short form loadTrans)
randomSeed=<int>                    Random seed (short form seed)
parallel=<int>                      Desired degree of parallelism in the data pipeline (short form n)
transform[<tag>]=<name>{<options>}  Transform (short form xf)
```

So yeah, we'll have to fix that. 😛 The number of places we call `GetFields` though to get the arguments attributes looks to be fairly small... but it is a complication I had not counted on. (Glad you mentioned it though.)

**Comment by mareklinka** [comment]
I honestly did not expect the N7 logo recognition over here, @TomFinley, but... awesome 😄 

I checked for `GetFields` and `GetField` usages in the solution and there is only about 10 of them in total. The only one I can safely identify as having to do with argument parsing is the one in [CmdParser](https://github.com/dotnet/machinelearning/blob/0c62e30b4d9eabb60322b2a3e75bc90e20007889/src/Microsoft.ML.Core/CommandLine/CmdParser.cs#L469). I might have missed something, though.

Also, should I also try and internalize other fields on argument classes where `Visibility == CmdLineOnly`?

**Comment by TomFinley** [comment]
Yeah. It's easy to forget now, but BioWare for a long while put out some amazing stuff. Aside from DA:O (also BioWare) I don't recall too many games that rival ME:1-3 in terms of narrative worldbuilding. ("Choose a color" aside, I guess.)

Sorry for the delay in response, I wanted to investigate the questions to be certain.

Regarding the more general rules of `CmdLineOnly`, I *might* think that is generally correct, since the only usages of that I am aware of is a case where there is a similar entry-point-only alternative, and usually the entry-point-only alternative is something also more appropriate for API usage, since the command line sometimes relies on establishing "mini-languages" that we don't really want to expose people to. (But, I'd hesitate without further investigation to call that a blanket policy.) Indeed these might be cases I'd imagine to be similar to the one issue uncovered by @stephentoub, and they should as you say probably likewise be fixed, probably.

Regarding `GetFields`, as you said, not too many, very easy to enumerate. I posted my impressions of each general "cluster" of cases below. ✔️ means should be changed to work over internal fields, ❌ means should probably be left alone and continue to work over just public fields.

1. https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.Core/CommandLine/CmdParser.cs#L469

   ✔️ This one should be made to work over `internal`, as already discussed.

2. https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.Core/CommandLine/CmdParser.cs#L1309

   ❌ This one *seems* to be enumerating the possible values for an `enum`. (Surely there are better ways, but, I guess this works?) We can leave it as is.

3. https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.Core/CommandLine/CmdParser.cs#L2221

   ❌ Analogous to prior case, "fine" as is.

4. https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.Core/EntryPoints/EntryPointUtils.cs#L53

   ✔️ Analogous for first case, but for the sake of enumerating these attributes for the sake of so called entry-points instead of command line, but same idea.

5. I see [two in `InputBuilder.cs`](https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.Data/EntryPoints/InputBuilder.cs#L61), but this is about . ❌

6. `JsonManifestUtils.cs`, three usages here.

   https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.EntryPoints/JsonUtils/JsonManifestUtils.cs#L157

   https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.EntryPoints/JsonUtils/JsonManifestUtils.cs#L279

   https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.EntryPoints/JsonUtils/JsonManifestUtils.cs#L479

   ✔️ Again analogous. I'm seeing a general rule, if we're inspecting it to see if it has this `ArgumentAttribute` or the entry-points output "dual" of this, `TlcModule.OutputAttribute`, then it should probably be changed to pay attention to internals.

7. https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/src/Microsoft.ML.LightGBM/LightGbmArguments.cs#L60

   ✔️/❌? In one sense, again analogous, but in addition to whatever is being done here they should *also* have restricted their attention to those things with the correct attribute on it, I think. (I don't like this code. It makes me nervous. If you wanted to improve it, that's fine, but it will continue to work as is, under the current circumstances. So maybe we just don't touch it, since it's so weird.)

8. https://github.com/dotnet/machinelearning/blob/bdc9a9e1348abfdd2287d43e2633089e9eba36c7/test/Microsoft.ML.Tests/CollectionsDataViewTest.cs#L184

   ❌ Unrelated test case.

I think that covers most of the cases.

**Comment by mareklinka** [comment]
Great analysis and guidance, thanks a lot. I will try and push the recommended changes tomorrow. I'll also try and have a look at the `CmdLineOnly` thing, see what happens. If we're not sure, I guess we can always spin the topic into a separate issue and get a more focused conversation going?

**Comment by mareklinka** [comment]
I took care of most of the discussed places. Regarding point 7 (the `LightGbmArguments`), I'll give it another look later and push it as a separate commit (if at all). From a cursory glance, you are most likely right - the code should check for the attribute, if for nothing else then to be consistent with other similar usages.

If we introduced this check, the current code should still work - looks like all the fields there have the attribute on them (I'll double-check this, probably write some smoke tests).
